packages/block-library/src/gallery/deprecated.js

```diff
@@ -19,6 +20,189 @@ import {
 } from './constants';
 
 const deprecated = [
+	{
+		attributes: {
+			images: {
+				type: 'array',
+				default: [],
+				source: 'query',
+				selector: '.blocks-gallery-item',
+				query: {
+					url: {
+						type: 'string',
+						source: 'attribute',
+						selector: 'img',
+						attribute: 'src',
+					},
+					fullUrl: {
+						type: 'string',
+						source: 'attribute',
+						selector: 'img',
+						attribute: 'data-full-url',
+					},
+					link: {
+						type: 'string',
+						source: 'attribute',
+						selector: 'img',
+						attribute: 'data-link',
+					},
+					alt: {
+						type: 'string',
+						source: 'attribute',
+						selector: 'img',
+						attribute: 'alt',
+						default: '',
+					},
+					id: {
+						type: 'string',
+						source: 'attribute',
+						selector: 'img',
+						attribute: 'data-id',
+					},
+					caption: {
+						type: 'string',
+						source: 'html',
+						selector: '.blocks-gallery-item__caption',
+					},
+				},
+			},
+			ids: {
+				type: 'array',
+				items: {
+					type: 'number',
+				},
+				default: [],
+			},
+			columns: {
+				type: 'number',
+				minimum: 1,
+				maximum: 8,
+			},
+			caption: {
+				type: 'string',
+				source: 'html',
+				selector: '.blocks-gallery-caption',
+			},
+			imageCrop: {
+				type: 'boolean',
+				default: true,
+			},
+			linkTo: {
+				type: 'string',
+			},
+			sizeSlug: {
+				type: 'string',
+				default: 'large',
+			},
+		},
+		supports: {
+			anchor: true,
+			align: true,
+		},
+		isEligible( { images } ) {
+			return (
+				images &&
+				images.some(
+					( image ) => image.alt !== __unstableStripHTML( image.alt )
+				)
+			);
+		},
+		migrate( attributes ) {
+			return {
+				...attributes,
+				images: attributes.images.map( ( image ) => {
+					let alt = image.alt;
+
+					if ( alt !== __unstableStripHTML( alt ) ) {
+						alt = __unstableStripHTML( image.alt );
+					}
+
+					return {
+						...image,
+						alt,
+					};
+				} ),
+			};
+		},
```

Comment:
Also, I'm not sure the migrate function is needed here because if I understand the PR properly, the attribute is not really rewritten, the same attribute will produce another output, that's it.

Reply:
Well the `alt` attribute -- which is implicit -- is being changed; so yeah, that attribute is part of the output (and not an explicit attribute in the block comment). Are you saying that a change like that doesn't need a migration? :thinking: If we don't add one, it means the block will load _with_ HTML in the `alt` attribute -- it will only be removed when saving it again, correct?

Reply:
yes, correct. but now that I think about it, there's something weird in the implementation of the gallery save function regardless.

If the image doesn't have any "alt", it will get assigned an "alt" text after saving because it will be parsed from the fallback. I wonder if we should remove the fallback personally and instead apply the fallback in `edit` function instead.

Reply:
> yes, correct. but now that I think about it, there's something weird in the implementation of the gallery save function regardless.
> 
> If the image doesn't have any "alt", it will get assigned an "alt" text after saving because it will be parsed from the fallback. I wonder if we should remove the fallback personally and instead apply the fallback in `edit` function instead.

You mean weird in terms of UX? So right now, after saving and reloading, the `alt`, if otherwise empty, would be prefilled with the `caption`. Moving to `edit` would essentially do the same though, no?

Whatever we do, let's please make sure to strip the HTML, and provide a working migration path. This [issue](https://github.com/WordPress/gutenberg/issues/26549) (when combined with some other HTML mangling/sanitization) is causing [havoc](https://github.com/Automattic/wp-calypso/issues/46813).

Reply:
I mean the block saving is not idempotent. If you save  the same post twice without changes, you get different attributes.

What if we do this:

 - When adding an image, use the caption as a fallback for the alt property. (and escape it)
 - Remove the fallback in save

Would that work?

Reply:
I think that makes sense, yeah 👍🏼

Reply:
Okay, that means basically going with what @WunderBart [originally had](https://github.com/WordPress/gutenberg/pull/26574#discussion_r514391807).

I'm still not entirely sure what that means for the blocks out there that now have HTML in their `alt` attributes :thinking: Apparently, we can't use a migration to fix them. So I guess we'd need to take care of that removal in (both?) `edit` and `save`? :thinking:

Reply:
> I mean the block saving is not idempotent. If you save the same post twice without changes, you get different attributes.
> 
> What if we do this:
> 
> * When adding an image, use the caption as a fallback for the alt property. (and escape it)
> * Remove the fallback in save
> 
> Would that work?

@WunderBart and I just had a call: it won't work exactly as you described, as the caption can be edited in the editor once the image has been inserted. So we'd basically need to update through the caption's `RichText` component's `onChange`. This probably means adding some additional criteria (only update the `alt` if it's empty, _or_ if `alt === __unstableStripHTML( caption )`).

![gallery](https://user-images.githubusercontent.com/96308/97870272-0a572700-1d13-11eb-87ff-48bb2f20145d.gif)

cc/ @youknowriad

Reply:
> This probably means adding some additional criteria (only update the alt if it's empty, or if alt === __unstableStripHTML( previousCaption )).

This makes sense yes.

Reply:
Thanks for helping with this, folks! 🙌 

> I'm still not entirely sure what that means for the blocks out there that now have HTML in their alt attributes 🤔 Apparently, we can't use a migration to fix them. So I guess we'd need to take care of that removal in (both?) edit and save?

IIUC, without the migration, we're in a situation in which a user needs to edit the block and save again in order to update the `image.alt` value. If that's true and we're OK with it, it should be sufficient to make an update on `save` only, like this:

https://github.com/WordPress/gutenberg/blob/a85b517c5993696f9dc77b747733a48778990113/packages/block-library/src/gallery/save.js#L44-L61

I tested this implementation for both scenarios - newly created `alt` and one that already existed and contained HTML. What do you think? Did I miss something here?

Reply:
Here's how I tested the above:

1. Checkout to master
2. Add gallery block
3. Add one image from the Media Library (make sure its `alt` attr is empty)
4. Add a caption to that image and add some HTML to it (make text bold, add link, etc.)
5. Save draft & reload the page
6. Check that image's source - its `alt` should contain HTML tags
7. Checkout this branch & refresh the page
8. Add another image to the same Gallery (also make sure its `alt` is empty)
9. Like previously, add a caption with some HTML in it
10. Save draft & refresh the page
11. Check source for both images - both `alt` attributes should be correctly stripped out of HTML.

Reply:
I know that @youknowriad was originally opposed to a `save`-based approach because of l[ack of idempotency](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515181556) (and suggested an [`edit`-based approach](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515977791) instead):

> If you save  the same post twice without changes, you get different attributes.

Although I'm not so sure anymore that's true?

- In a gallery block, insert an image without an `alt` text.
- Add a caption.
- Save the post. The caption is used for the `alt` attribute.
- Save again. Now that we already have an `alt` attribute, nothing changes.

@youknowriad Could you weigh in here?

Reply:
I guess some of the problem here is due to the fact that the `alt` attribute, if changed from the media modal, is propagated to the `alt_text` property of the image (whereas the same isn't true if set from the editor during `save`).

Since I would really like the issue with HTML in the `alt` attribute (which is [breaking](https://github.com/Automattic/wp-calypso/issues/46813) a lot of our users' galleries) to be fixed in 9.4, I propose that in case of doubt, we settle for a simple solution that we can all agree on and that we know will fix the issue, without introducing any new regressions; and that we figure out a 'better' solution later. I think that @WunderBart's current, `save`-based solution, fulfills that criterion ([see](https://github.com/WordPress/gutenberg/pull/26574#discussion_r516010096)).

Reply:
> I know that @youknowriad was originally opposed to a `save`-based approach because of l[ack of idempotency](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515181556) (and suggested an [`edit`-based approach](https://github.com/WordPress/gutenberg/pull/26574#discussion_r515977791) instead):
> 
> > If you save  the same post twice without changes, you get different attributes.
> 
> Although I'm not so sure anymore that's true?

Ah, good point! It's idempotent for newly created `alt`s only. For `alt`s that already exist and contain `HTML` it will produce a different output this one time (migration-like). Can we live with that, @youknowriad?

Reply:
I'll add that I believe that we need some logic like this in the `save` anyway, for the purpose of the pseudo-migration for `alt` attributes that were touched with GB 9.2 active, and had HTML inserted. (I don't think we can solve that solely at `edit` level.) And since the change (at `save`) level is sufficient also for newly created galleries, there's nothing left to do at `edit` level.

Reply:
> I'll add that I believe that we need some logic like this in the save anyway, for the purpose of the pseudo-migration for alt attributes that were touched with GB 9.2 active, and had HTML inserted.

For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

How many persons are affected by this issue and for Core, it's still a development plugin.

Reply:
I'd also note that in all cases, the user has to open and save the post anyway since migrations don't run automtically.

Reply:
> For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

What if we "mapped" the `alt` prop in `gallery.image.js`'s `compose` instead?

```js
export default compose( [
	withSelect( ( select, ownProps ) => {
		const { getMedia } = select( 'core' );
		const { id } = ownProps;
		const image = id ? getMedia( parseInt( id, 10 ) ) : null;

		if ( image ) {
			let alt = image.alt_text || image.alt || image.caption;

			if ( alt && alt !== __unstableStripHTML( alt ) ) {
				alt = __unstableStripHTML( alt );
			}

			image.alt = alt;
		}

		return {
			image,
		};
	// ...
```

This would handle `alt` fallback but the user would see what they're saving this way. It still requires setting `alt` on caption edit but that's not a problem.

Reply:
> For me personally, I don't see that as a big issue. I see the conceptual issue as more important: (the conceptual issue being the save function not idempotent creating different block attributes without touching the editor).

I'm all for purity of code and all that, but it's the impact of this issues that worries me: Post content actually gets broken, so people will have to revert to an earlier revision (and might lose the changes they were working on). OTOH, the feature that was added by the [PR](https://github.com/WordPress/gutenberg/pull/26082) that caused this is an (arguably minor) a11y improvement. IMO, those aren't strong enough reasons to keep around a feature that has such negative side effects.

> How many persons are affected by this issue and for Core, it's still a development plugin.

Please refer to https://github.com/Automattic/wp-calypso/issues/46813. I'm counting ~15 so far (some of those comments refer to the WP.com forums, where each thread can potentially contain multiple reports. Here's a few:

https://wordpress.com/forums/topic/gallery-block-problem/
https://wordpress.com/forums/topic/galleries-not-showing-up-right/
https://wordpress.com/forums/topic/unreadable-appearance-of-gallery-page-after-editing/
https://wordpress.com/forums/topic/cant-make-changes-to-website-and-drop-down-menus-are-not-working/
https://wordpress.com/forums/topic/galleries-not-showing-up-right/

> I'd also note that in all cases, the user has to open and save the post anyway since migrations don't run automtically.

That's correct, but that's how migrations work, isn't it? This could be addressed by adding additional tooling to e.g. 'fix' the database upon plugin upgrade, but let's not get the lack of a perfect solution in the way of a sufficient one.

Needing to open and save can at least help with the cases where users with GB 9.2 were editing their galleries and because of the issue lost content, revert the post to a previous revision, and would like to re-apply those changes.

---

Anyway, is the compromise to simply remove the `image.caption` fallback for `alt`? I'd be okay with that, for the sake of at least preventing any further breakage into 9.3, even if it means that we won't unbreak existing content.

Reply:
> Anyway, is the compromise to simply remove the `image.caption` fallback for `alt`? I'd be okay with that, for the sake of at least preventing any further breakage into 9.3, even if it means that we won't unbreak existing content.

Filed the alternative PR: #26676